shared/src/types.rs

```diff
@@ -428,99 +432,182 @@ impl Display for Peer {
     }
 }
 
-#[derive(Debug, PartialEq)]
-pub struct PeerDiff {
-    pub public_key: String,
-    pub endpoint: Option<SocketAddr>,
-    pub persistent_keepalive_interval: Option<u16>,
-    pub is_disabled: bool,
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChangeString {
+    name: &'static str,
+    old: Option<String>,
+    new: Option<String>,
 }
 
-impl Peer {
-    pub fn diff(&self, peer: &PeerConfig) -> Option<PeerDiff> {
-        assert_eq!(self.public_key, peer.public_key.to_base64());
-
-        let endpoint_diff = if let Some(ref endpoint) = self.endpoint {
-            match endpoint.resolve() {
-                Ok(resolved) if Some(resolved) != peer.endpoint => Some(resolved),
-                _ => None,
-            }
-        } else {
-            None
-        };
-
-        let keepalive_diff =
-            if peer.persistent_keepalive_interval != self.persistent_keepalive_interval {
-                self.persistent_keepalive_interval
-            } else {
-                None
-            };
+impl Display for ChangeString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}: {} => {}",
+            self.name,
+            self.old.as_deref().unwrap_or("[none]"),
+            self.new.as_deref().unwrap_or("[none]")
+        )
+    }
+}
 
-        if endpoint_diff.is_none() && keepalive_diff.is_none() {
-            None
-        } else {
-            Some(PeerDiff {
-                public_key: self.public_key.clone(),
-                endpoint: endpoint_diff,
-                persistent_keepalive_interval: keepalive_diff,
-                is_disabled: self.is_disabled,
-            })
+impl ChangeString {
+    pub fn new<T, U>(name: &'static str, old: Option<T>, new: Option<U>) -> Self
```

Comment:
Sorry, I'm not sure what you mean here. Can you show in code what you mean?

Reply:
The codebase has:
```rust
pub struct ChangeString {
    ...
}

impl Display for ChangeString {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result { ... }
}

impl ChangeString {
    pub fn new<T, U>(name: &'static str, old: Option<T>, new: Option<U>) -> Self { ... }
}
```

I would have written:
```rust
pub struct ChangeString {
    ...
}

impl ChangeString {
    pub fn new<T, U>(name: &'static str, old: Option<T>, new: Option<U>) -> Self { ... }
}

impl Display for ChangeString {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result { ... }
}
```

I guess I was writing my comment when I was in extra curious mood, it is kinda unimportant. :)

Reply:
Ah! Yes, I do tend to add the `impl Struct` at the end, just out of habit. I guess in my head it makes more sense since it's closer to the `#[derive(Clone, Debug)]` declaration which is also in essence a `impl X for Struct`